enhancements/security/manual-cloud-credentials.md

```diff
@@ -0,0 +1,170 @@
+---
+title: manual-cloud-credentials
+authors:
+  - "@dgoodwin"
+reviewers:
+  - @joelddiaz
+  - @smarterclayton
+  - @abhinavdahiya
+  - @derekwaynecarr
+  - @soltysh
+approvers:
+  - @joelddiaz
+  - @abhinavdahiya
+  - @soltysh
+creation-date: 2020-05-14
+last-updated: 2020-05-14
+status: provisional
+---
+
+# Support Manual Provisioning of Cloud Credentials
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This proposal outlines a plan for fully supporting the pre-existing process for disabling the cloud-credential-operator (still running, but in a no-op mode) and manually provisioning the cloud credentials for your cluster. It allows users to avoid broken upgrades when credentials have changed, and provides them with the tools to detect and reconcile the problem.
+
+## Motivation
+
+The current best practice for installing OpenShift 4 involves providing openshift-install with `root` cloud credentials powerful enough to mint additional fine grained credentials for each component in the cluster that needs them. The root credentials are stored in a secret in the cluster. This mechanism allows each component to have only the permissions it needs, and the required permissions are defined in the release image and automatically upgraded with the cluster itself.
+
+However some customers are unable or unwilling to do this and would prefer to disable automatic credential provisioning, and instead manage it themselves. It is possible to disable the cloud-credential-operator today by injecting a `ConfigMap` into the openshift-install manifests directory, along with several `Secrets` containing the cloud provider credentials the user manually configured. While this process exists today, it is not officially supported as we believe there are situations that could arise during upgrade where credentials have changed, and the upgrade will block and go into a failed state, with little visibility as to why or how it can be fixed.
+
+### Goals
+
+List the specific goals of the proposal. How will we know that this has succeeded?
+
+Goals for users who have chosen to disable the cloud credential operator:
+
+  * Users can easily mint and inject their own cloud credential secrets at install time.
+  * Users will be prevented from upgrading *if* credentials have changed in an available release image update.
+  * Users can upgrade without failures when the new release image contains credential changes.
+  * Users have tooling at their disposal to aid them in detecting what credentials changes have occurred in a release image.
+  * Users have tooling at their disposal to aid them in updating their credentials with new permissions.
```

Comment:
hmm.. i don't know if this makes sense based on the fact that users don't want to use cred-minter in the first place. Are we assuming they will use our process now?

Reply:
Do they not want to use tooling and do it all by hand? Or do they not want admin creds floating around in their cluster? My focus has been the latter, I can't imagine anyone would actually want to parse a release image and go line by line updating credentials in AWS if they didn't have to. I would want that tooling available so I could run it locally keeping my admin creds out of the cluster, and display exactly what it's doing.

Naturally tooling could be optional and if you really did want to go manually do everything in AWS, you can do that, and just flag that you've done it in the same manner the CLI tooling would.

 Customer we talked to last week was not interested in manual processes, maybe some want that, but I will pose the question to PM to see if CLI tooling to let people do this automatically but without root creds in the cluster is something customers want.

Reply:
I believe there's a couple reasons for not wanting to use CCO. The most obvious reason (which was already mentioned) is not wanting to leave their long-term, admin credentials lying around in the cluster. In the case of many of the public sector customers, they take this one step further and only enable credentials for select services rather than every cloud-enabled service in the cluster. 

The other popular reason stems from the lack of a private IAM endpoint for disconnected users. They don't want to (or aren't allowed to) enable proxy connectivity in these environments, so they must manually provide credentials if they're going to be disabling the CCO. 

Finally, the other reason we've heard quite a bit is somewhat tangential to disable CCO but rather moving to short-term, temporary credentials that can be renewed/rotated on a given interval. This may eliminate some of the previously mentioned reasons for disabling CCO if they can use a service like STS where it can still be used in an air-gapped environment.

Reply:
Ok so I will continue with assumption that our primary goal here is keeping admin creds out of the cluster.

@kedube would you agree that (1) CLI tooling to run locally to automate cred provisioning is useful (I'm guessing yes) and (2) is this tooling required for 4.6 or could we still "launch" this as supported and documented without said tooling.

Reply:
(1) Yes. (2) I believe so. Obviously tooling makes it easier, but I don't think it should be blocker to use even if the experience isn't the best due to lack of tooling. Otherwise, I'm sure tooling will be something customers request! ;)

Reply:
My initial reaction to any tooling was: if I prefer not to allow an operator to mint creds, how is that really much different than some additional automated tool that mints creds? Both would be open source. Seems to me the only difference is that admin credentials aren't lying around in the cluster if some CLI tool is used.

Reply:
That is the key difference, the admin cred is never stored in the cluster with this approach. Personally I understand not wanting to store an admin cred in your cluster, and I'd be much more comfortable using a ~/.aws/credentials on my local workstation periodically to reconcile my creds prior to upgrade.